File: windows-client/PrintixSend/Services/Logger.cs
using System;
using System.IO;

namespace PrintixSend.Services;

public class Logger
{
    private readonly string _logFile;
    private readonly object _lock = new();

    public Logger()
    {
        var baseDir = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "PrintixSend", "logs");
        Directory.CreateDirectory(baseDir);
        _logFile = Path.Combine(baseDir, $"printix-send-{DateTime.Now:yyyyMMdd}.log");
    }

    public string LogFilePath => _logFile;

    public void Info(string msg)  => Write("INFO",  msg);
    public void Warn(string msg)  => Write("WARN",  msg);
    public void Error(string msg) => Write("ERROR", msg);
    public void Error(string msg, Exception ex) => Write("ERROR", $"{msg} — {ex.GetType().Name}: {ex.Message}\n{ex.StackTrace}");

    private void Write(string level, string msg)
    {
        var line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} [{level}] {msg}";
        lock (_lock)
        {
            try { File.AppendAllText(_logFile, line + Environment.NewLine); }
            catch { /* ignore */ }
        }
#if DEBUG
        System.Diagnostics.Debug.WriteLine(line);
#endif
    }
}

File: macos-client/Sources/PrintixSendCore/Logger.swift
import Foundation

// Einfacher Datei-Logger nach dem Muster des Windows-Clients.
// Schreibt nach ~/Library/Logs/PrintixSend/printix-send-YYYYMMDD.log
// und – optional – zusätzlich nach stderr, damit die CLI auch im Terminal
// Feedback gibt, wenn man sie manuell aufruft.

public final class AppLogger: @unchecked Sendable {
    public static let shared = AppLogger()

    private let queue = DispatchQueue(label: "de.printix.send.log")
    private let formatter: DateFormatter
    private let dayFormatter: DateFormatter
    public var alsoStderr: Bool = false

    private init() {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        self.formatter = f
        let d = DateFormatter()
        d.locale = Locale(identifier: "en_US_POSIX")
        d.dateFormat = "yyyyMMdd"
        self.dayFormatter = d
    }

    private var logDir: URL {
        let home = FileManager.default.homeDirectoryForCurrentUser
        return home.appendingPathComponent("Library/Logs/PrintixSend", isDirectory: true)
    }

    private var logFile: URL {
        let name = "printix-send-\(dayFormatter.string(from: Date())).log"
        return logDir.appendingPathComponent(name)
    }

    private func ensureDir() {
        try? FileManager.default.createDirectory(at: logDir, withIntermediateDirectories: true)
    }

    public func info(_ msg: String) { write("INFO",  msg) }
    public func warn(_ msg: String) { write("WARN",  msg) }
    public func error(_ msg: String) { write("ERROR", msg) }

    private func write(_ level: String, _ msg: String) {
        let line = "\(formatter.string(from: Date())) [\(level)] \(msg)\n"
        queue.async { [self] in
            ensureDir()
            if let data = line.data(using: .utf8) {
                if let fh = try? FileHandle(forWritingTo: logFile) {
                    fh.seekToEndOfFile()
                    fh.write(data)
                    try? fh.close()
                } else {
                    try? data.write(to: logFile)
                }
            }
            if alsoStderr {
                FileHandle.standardError.write(Data(line.utf8))
            }
        }
    }
}
